Initial Classes

XedinUnknown opened the pull request:
- PHP evaluator abstracts the source of PHP, and can evaluate PHP code given a context (variables). File-based evaluator included, as well as a factory.
- PHP evaluator-based template abstacts templating. It allows specifying a default context, as well as functions. The context and functions will be made available to the evaluated PHP in the form of variables `$c` and `$f` respectively. This allows the template body to only have to use what is provided, instead of relying on anything outside.
- Exceptions.

XedinUnknown committed:
Now depending on interface package

XedinUnknown committed:
Added dev deps for testing

XedinUnknown committed:
Added dep on abstract package

XedinUnknown committed:
Updated dependencies

XedinUnknown committed:
Switched to PHP 7.0

XedinUnknown committed:
Added Docker environment

Currently facilitates a PHP interpreter + Composer.

XedinUnknown committed:
Got rid of XPMock

XedinUnknown committed:
Switched to PHPUnit 6

XedinUnknown committed:
No longer depending on abstract package

XedinUnknown committed:
Depending on latest interface package

XedinUnknown committed:
Now depending on vfsStream package for testing file access

XedinUnknown committed:
Depend on PSR-11 package for testing

XedinUnknown committed:
Remove dependency on php-vfs package

No longer needed for testing. Using vfsstream instead.

XedinUnknown committed:
Added `PhpFileEvaluator`

XedinUnknown committed:
Dependencies updated

Can now throw `Exception` from templates while rendering.
Without this, the i18n of rendering exception messages would break the standard.

XedinUnknown committed:
Added `PhpTemplate`

XedinUnknown committed:
Added exceptions

XedinUnknown committed:
Added `FilePhpEvaluatorFactory`

XedinUnknown committed:
CI will now test on PHP 7 and above, and not below

XedinUnknown committed:
PHPStorm configs updated

XedinUnknown committed:
Added missing PHPDocs, and corrected EOF

mecha commented on src/PhpEvaluator/FilePhpEvaluator.php:
```diff
@@ -0,0 +1,60 @@
+use Exception;
+
+/**
+ * A PHP evaluator that directly evaluates PHP files via `include`.
```
Actually it looks like you are using `require`

XedinUnknown replied:
Yea, but for PHP `require` and `include` are the same thing, except for what happens if the file isn't there. Still, maybe better use `include` in the actual code. What do you think?

mecha replied:
I'm biased towards `require` as I'd prefer my applications to stop execution if they encounter an error, rather than continue "partially" running. They either jog till the very end or not at all. No limping allowed.

But that's just my opinion. I can understand if others disagree since we're dealing with output here, which tends to not be mission critical.

XedinUnknown replied:
Yes, I understand, and I feel the same way. One of my points was that `include` and `require` share the same mechanic. But yea.

mecha replied:
Yeah. But the class doc still needs to match what is actually used.

XedinUnknown replied:
Fixed. Now using `include`.

mecha commented on src/PhpEvaluator/FilePhpEvaluator.php:
```diff
@@ -0,0 +1,60 @@
+        $fn->bindTo(null);
+
+        return $fn;
+    }
```
The loaded file will have access to `$this`, from which it could potentially call `evaluate()` and `_isolateFileScope()`.

If you don't want this, then maybe this method ought to be converted into a generic function, in a file that is loaded via a file map from Composer.

mecha replied:
This actually gives me an idea ...

XedinUnknown replied:
Why would it have access to `$this` if I am explicitly unbinding it on line 56?

mecha replied:
Completely missed that line! My bad.

mecha commented on src/Template/PhpTemplate.php:
```diff
@@ -0,0 +1,232 @@
+    {
+        return function (string $funcName, ...$args) use ($context) {
+            return $this->invokeFunction($funcName, $args);
+        };
```
`$context` is not used.

XedinUnknown replied:
Yep, it's true. However, I thought it was a good idea to pass it to all methods that create template variables, because they may want to depend on context.

What would you suggest in this case? Just remove it from the `use`?

mecha replied:
I'd remove the `use`. If a derived class wants to use it, the parameter is there. But in this case I wouldn't add any code that does nothing, purely for clarity and ease of mind.

XedinUnknown replied:
Removed `use`.

mecha commented on src/Template/PhpTemplate.php:
```diff
@@ -0,0 +1,232 @@
+    /**
+     * @var array<string, callable>
+     */
+    private $functions;
```
What is this blasphemy?

XedinUnknown replied:
Just settings of PHPStorm auto-generation. Will fix, thanks :)

mecha commented on src/Template/PhpTemplate.php:
```diff
@@ -0,0 +1,232 @@
+     */
+    protected function getContextValue($context, string $key, $default = null)
+    {
+        try {
+            $this->validateContext($context);
+        } catch (InvalidArgumentException $e) {
+            throw new InvalidArgumentException($this->__('Could not retrieve value for key "%1$s": invalid context', $key), 0, $e);
+        }
```
If the context is not allowed to change during rendering, then I suggest validating the context only once, before rendering begins. It allows potential context errors to be reported prior to rendering, as well as being reported even in cases where the template does not use the context.

The latter could be especially beneficial for the developer, in cases where for example an invalid context is given to a template that does not use the context, but some third-party overrides the template, say through some module system :wink:, and the overriding template _does_ use the context. That would trigger this exception, but the original author would have never reproduced it.

XedinUnknown replied:
The reason I validate it here is that it is only this part of the code that actually depends on the context being a specific thing. So, I don't really see how this method could work without this validation: if it is not one of the supported types, this method must throw.

I do agree, however, that it could be a good idea to validate the context before rendering begins. I would need to also allow it to be `null`, in addition to the supported types, in that particular case.

mecha replied:
Sounds good to me.

XedinUnknown replied:
Fixed. Now validating.

mecha commented on src/Template/PhpTemplate.php:
```diff
@@ -0,0 +1,232 @@
+
+        // It's an array-like
+        {
+            if (!array_key_exists($key, $context)) {
```
Be advised that `array_key_exists` on objects, even if they implement `ArrayAccess`, is [deprecated as of PHP 7.4](https://3v4l.org/iRHCK).

XedinUnknown replied:
That is [exactly right][array_key_exists_with_objects]. In fact, apparently `array_key_exists()` _never_ actually supported `ArrayAccess`. Looks like I have to add explicit support.

[array_key_exists_with_objects]: https://wiki.php.net/rfc/deprecations_php_7_4#array_key_exists_with_objects

XedinUnknown replied:
I added tests for this via data provider, but I am unable to reproduce this issue due to the fact that PHPUnit 6 does not support PHP 7.4, and gives an error before the above code can run.

XedinUnknown replied:
Here's what's strange, though: If `array_key_exists()` has no support for `ArrayAccess`, and "only works on mangled properties", then why does this code work even on PHP 7.3 and below?

mecha replied:
[Because public properties are not name-mangled](https://3v4l.org/jqO9a), and in order for `$context` to be useful in this context (pun definitely intended) the properties need to be public.

Unless they are magic, in which case `array_key_exists` will definitely not work.

XedinUnknown replied:
How does the fact that `array_key_exists()` seems to work on `ArrayAccess` relate to public properties and their names?

mecha replied:
Not sure how to answer that. I simply replied to your question.

> If array_key_exists() has no support for ArrayAccess, and "only works on mangled properties", then why does this code work even on PHP 7.3 and below?

The code works on PHP 7.3 and below because:

* It [does work](https://3v4l.org/ufn9L) for `ArrayAccess`
* It [does work](https://3v4l.org/cA5XI) for `object` (as long as the properties are public, because their mangled names would be the same as the original names)

XedinUnknown replied:
I get that it works, but this seems to contradict the documentation.

XedinUnknown replied:
Anyhow, fixed.

mecha commented on src/PhpEvaluator/FilePhpEvaluator.php:
```diff
@@ -0,0 +1,60 @@
+    /**
+     * @var string
+     */
+    private $filePath;
```
Why `private`?

XedinUnknown replied:
PHPStorm settings. Will change.

XedinUnknown committed:
Fixed `private` properties.

XedinUnknown committed:
Now testing also on PHP 7.4

XedinUnknown committed:
Now allowing PHP nightly builds to fail, and fast finish

XedinUnknown committed:
Change CI distro to `trusty`

This should allow PHP 7.4 to be installed.

XedinUnknown committed:
Allow build failures on PHP 7.4

It appears that PHPUnit 6 does not support 7.4.

XedinUnknown commented:
So, I tried to include PHP 7.4 in the build. However, this is not possible: PHPUnit 6, which is the highest version of PHPUnit that supports PHP 7.0, does not support PHP 7.4. This results in PHPUnit itself throwing errors like this one:

> 1\) Dhii\Output\FuncTest\Template\PhpTemplateTest::testRender
Function ReflectionType::__toString() is deprecated
/home/travis/build/Dhii/php-template/test/functional/Template/PhpTemplateTest.php:38
/home/travis/build/Dhii/php-template/test/functional/Template/PhpTemplateTest.php:66

That's simply the place where I call `PHPUnit_Framework_MockObject_MockBuilder::getMock()`.

For this reason, PHP 7.4 is included in the build, but is allowed to fail. Thus, it can be used for information.

XedinUnknown committed:
Now testing rendering with all possible types of context

XedinUnknown committed:
Removed NetBeans project directory

mecha commented:
@XedinUnknown Can't you change the build's `before_script` to upgrade PHPUnit if the PHP version is `7.4` or later?

mecha commented:
I ran a quick test with an empty project using the below constraint for PHPUnit, and `dev` minimum stability:

```
"phpunit/phpunit": "^6.0 || ^7.0 || ^8.0"
```

Using the `config.platform.php` property in `composer.json`, I tried installing on PHP 7.0, 7.1, 7.2, 7.3 and 7.4 to see which versions of PHPUnit Composer would install. These are my results:

| PHP | PHPUnit   |
|-----|-----------|
| 7.0 | 6.5.14    |
| 7.1 | 7.5.x-dev |
| 7.2 | 8.5.x-dev |
| 7.3 | 8.5.x-dev |
| 7.4 | 8.5.x-dev |

I'm not sure if v7 and v8 are BC in terms of API with v6. But can't hurt to try, eh?

XedinUnknown committed:
Use named Docker containers

This allows avoiding clashes with other projects

XedinUnknown committed:
Upgraded version of Docker compose config

Now matches that used by PHPStorm.

XedinUnknown committed:
Allow PHPUnit 7 and 8

This is for CI to be able to test on higher versions of PHP correctly

XedinUnknown commented:
For some reason, it keeps installing PHP 6. See [build log][].

[build log]: https://travis-ci.com/Dhii/php-template/jobs/273915936

mecha commented:
I think I know what the problem is.

I took the same test project, pasted the dev-dependencies from this package and did `composer update`. Composer downgraded PHPUnit to `6.2.4`.

My recommendation is to modify the TravisCI build to install PHPUnit globally. This should ensure that the version of PHPUnit that is installed in the image works for that image's PHP version.

```
before_script:
  - composer update --prefer-dist
  - composer global require phpunit/phpunit
  - export PATH=~/.composer/vendor/bin:$PATH
script:
  - phpunit
```

XedinUnknown committed:
Remove deprecated code coverage reporting method

XedinUnknown commented:
So, you're saying that other dev dependencies rely on specifically PHPUnit `^6.0` or something like that, I guess?

XedinUnknown committed:
Remove useless PHP CS standard package

XedinUnknown committed:
Disallow PHPUnit 8

XedinUnknown committed:
Add specific support for `ArrayAccess` contexts

`array_key_exists()` does not work on those, and causes a deprecation
warning in PHP 7.4+.

XedinUnknown committed:
Do not allow PHP 7.4 builds to fail any longer

mecha commented:
Yeah. Looks like that did the trick.

XedinUnknown committed:
Now using `include` instead of `require`

XedinUnknown committed:
Removed unused variable import

XedinUnknown committed:
Now validating context before rendering begins
